File: Algorithms/Algos_in_JS/Sorting/quicksort.js
function partition(array, start, end) {
  let pivot = array[end];
  let i = start - 1;
  for (let j = start; j <= end - 1; j++) {
    if (array[j] < pivot) {
      i++;
      let temp = array[j];
      array[j] = array[i];
      array[i] = temp;
    }
  }
  i++;
  let temp = array[i];
  array[i] = pivot;
  array[end] = temp;
  return i;
}

function sort(array, start, end) {
  if (start < end) {
    let pivotIndex = partition(array, start, end);
    sort(array, start, pivotIndex - 1); // Recursively sort the left subarray
    sort(array, pivotIndex + 1, end); // Recursively sort the right subarray
  }
}

function quickSort(array) {
  sort(array, 0, array.length - 1);
}

const array = [5, 3, 1, 4, 2];
console.log("Original Array:", array);
quickSort(array);
console.log("Sorted Array:", array);

File: algorithms/algos-in-cpp/sorting-searching/quicksort.c
#include <stdio.h>
#include <stdlib.h>

void printArray(int *A, int n)
{
    for (int i = 0; i < n; i++)
    {
        printf("%d ", A[i]);
    }
}

int partition(int A[], int low, int high)
{
    int pivot = A[low];
    int i = low + 1;
    int j = high;
    int temp;

    do
    {

        while (A[i] <= pivot)
        {
            i++;
        }

        while (A[j] > pivot)
        {
            j--;
        }

        if (i < j)
        {
        // Swap low, high
            temp = A[i];
            A[i] = A[j];
            A[j] = temp;
        }

    } while (i < j);

    // Swap pivot with high
    temp = A[low];
    A[low] = A[j];
    A[j] = temp;

    return j;
}

void quickSort(int A[], int low, int high)
{
    int partitionIndex;

    if (low < high)
    {
        partitionIndex = partition(A, low, high);
        quickSort(A, low, partitionIndex - 1);  // sort left partition
        quickSort(A, partitionIndex + 1, high); // sort right partition
    }
}

int main()
{
    int array[] = {2, 1, 5, 4, 3};
    int n = 5;

    printf("\nBefore Quick Sort\n");
    printArray(array, n);

    quickSort(array, 0, n - 1);

    printf("\nAfter Quick Sort\n");
    printArray(array, n);

    printf("\n");

    return 0;
}